### ISSUE TITLE
Create Service Folder Again

### ISSUE THREAD

**Issue by ac360** [serverless/serverless#1726]
##### Serverless Framework Version:  
v.1-alpha-2

##### Operating System: 
osx 10.11.12

##### Expected Behavior:
Expected Serverless to create a folder in the CWD and put scaffolding in that folder.

##### Actual Behavior:
Serverless puts files in CWD.

##### Comments:
Why was this change made?  Why do users have to manually create folders for services?  Many other frameworks auto-create folders for you.  This creates extra steps for the user.

Also, why is the service name no longer auto-populated?  I immediately deployed something with the incorrect service name, now I have to clean it up.  These new changes result in a worse out-of-the-box experience :(

**Comment by eahefnawy** [comment]
After some discussion with @mthenw and @flomotlik it seems that it's actually more common to generate boilerplate in the CWD (Apex & Meteor are doing it this way for instance). It skips the step of `cd` into the folder after creation, and gives the user control over the folder name. I guess it's pretty opinionated though, so I'm fine with both.

We dropped the name option because we want to rely on templates without having to process them at all (just copy the entire thing). Processing `serverless.yml` for example to add the name will cause yaml features to disappear, like comments and anchors...etc ... and we wanted to keep the comments in particular.

**Comment by ac360** [comment]
Frameworks that create folders for their scaffolding:

* Meteor - https://www.meteor.com/tutorials/blaze/creating-an-app
* Ruby On Rails - http://guides.rubyonrails.org/getting_started.html
* Django - https://docs.djangoproject.com/en/1.10/intro/tutorial01/#creating-a-project
* Revel - https://revel.github.io/tutorial/createapp.html

The extra command to make the directory is unnecessary.  The Framework should always reduce time to action.

Let's figure out how to auto-generate the service name too.  What if we remove the `service` property from the templates, and simply prepend it to line 1, when copying the template over.  Perhaps that will prevent us from having to do any disruptive modifications to the other parts.  Comments will exist below the service name, but that's not a big deal.  Also, all of the beginning comments in the `aws-nodejs` template could mostly be removed.  They are obvious and don't add much value. 

I'm being a stickler about this because the overall V.1 on-boarding experience is getting increasingly worse than V.0:

* V.0 created credentials for the user.  V.1 makes the user do this on their own.  Newcomers to AWS have no idea how to do this.
* V.0 had select menus w/ templates for runtimes.  V.1 the user has to look up the templates in the docs, causing them to immediately step away from the CLI.
* V.0 created folders for the user.  V.1 makes the user create their own folder for some reason.
* V.0 named the user's scaffolding.  V.1 makes the user go in and manually edit the files before they can do anything.

**Comment by flomotlik** [comment]
Imho we should create a --path option for the create plugin as well where instead of creating it in the local directory you can create the project in a specific path.

Also agree on the service name, imho having a nice solution where people can set the name easily and we overwrite it would be cool. Maybe something where we do a simple string replacement. We took it out as we had trouble with the comments and wanted to get this out the door first.

**Comment by mthenw** [comment]
> Frameworks that create folders for their scaffolding:

I think it's relic from the past. 

People have different workflows. Some of them create repo first. Some of them create code first. There is many ways to achieve the same result. IMHO creating folder is confusing and framework should now automate so basic thing. 

I should be able to do 

```
serverless create
serverless deploy
```

not (I assume that we don't want to change WD because that would be super confusing)

```
serverless create
cd project
serverless deploy
```

**Comment by ac360** [comment]
Thanks for the input, @mthenw.

You left out the first part:

```
mkdir my-first-service && cd my-first-service
serverless create
serverless deploy
```
vs.
```
serverless create
cd service
serverless deploy
```

One way to think about it is to pretend you are doing a demo.  I really don't want to deal with the first option when giving a demo.  Making the directory first and cd'ing into it will often be forgotten.  I'll create files in a folder where I didn't want them and lose time cleaning that up (I already did this).  Whereas if I forget `cd service` it will simply error on deploy and say `No service found`.

Further, when we allow for installable services, will this also mean the user has to make a folder before they can install service files into it, since that's the pattern we've established?  That's even more unusual (e.g., downloading a git zip, any zip, or installing an npm module).

Let's change this back and focus on solving problems rather than diverging from V.0 without strong reason.  Again, the overall V.1 on-boarding experience is more complex than V.0.  That is not the goal of V.1.  Let's only reduce steps.

**Comment by pmuens** [comment]
TBH I'm also more comfortable when Serverless creates the directory for me and puts everything in there. I was really lost when we changed that the other day.

We should extend the `create` command with some options like this:

`serverless create --template aws-nodejs --path ./ --name my-awesome-service`

That could solve some of the main pain points. Some users have already requested the `--name` option because you have to rename your service anyway so why not directly passing it into the CLI?

**Comment by flomotlik** [comment]
Yup fully agree on the --name and --path option.

From a Workflow perspective I've found that creating subdirectories to be a lot more painful than the other way around typically, especially when I create projects regularly.

We want people to create a lot of services regularly, the typical use case (at least for me) is creating a Github repo, cloning it and then creating the files in that folder. I've always hated when Rails for example created the subfolder automatically for me (and there is an option to disable it)

The main reason we implemented it like this though is to move faster on the template system as we would otherwise have to do something with how we handle naming in the template files and still keep the comments in there (which are very important). I'm all for adding this back at some point, but imho compared to other things this has relatively low priority as it is very easily covered in the onboarding docs and I don't think something that people get blocked by (there are many more things that we need to work on that block people).

Regarding external services and how they are installed we have not decided yet. This is something we have to discuss in the future, but I don't think it has a particular impact on this issue specifically as services will be completely separately handled from how we create a new serverless app. We have to think about it and discuss it once we focus on how we want people to use and install services

**Comment by jagi** [comment]
I don't know if you experience the same problem but after updating to `1.0.0-beta.1.1` and after calling `serverless create --template aws-nodejs` it does not create any files in the current directory.

**Comment by pmuens** [comment]
Will be implemented in #1937

**Comment by ac360** [comment]
@flomotlik @pmuens @DavidWells Sorry for re-open, but I want to put in one last request for the CLI to make a subdirectory, like it used to. 

Both `mkdir` and `--path` make the on-boarding look complex.  They are lengthy CLI commands that look intimidating in our quick-start and on the front page of this website.  Now, there are requests to hide `mkdir` from the quick-start and website, which would mean we provide an incomplete example.

This shouldn't have been changed.  It has only added more on-boarding friction, which goes against our goals right now.  Is there a chance we could just make a subdirectory again?

Further, `name` should be a parameter again and we should auto-name the service, like V.0.  I have suggested a quick fix for the naming template problem to @eahefnawy.  This is not a hard problem and it should not disrupt the on-boarding path.  Please let me know where I can write down a suggested solution.

**Comment by pmuens** [comment]
@ac360 Have you looked into #1937 which was opened yesterday? It implements the `--path` idea (which feels natural to me when using the CLI as a developer).

There's already a WIP PR for the `name` option (https://github.com/serverless/serverless/pull/1910).

**Comment by ac360** [comment]
Ok, just discussed this with @pmuens.  It seems like we can accomplish everyone's use-cases with this suggestion:

```
npm i -g serverless
serverless create -p myService -t aws-nodejs
cd myService
serverless deploy
```

* Without `-p`, the service is created in the CWD, with a random service name.
* With `-p` the service is created in the specified path.  We'll strip the folder name from the path, and use that in the `serverless.yml` `name` property.  Also users will be able to do `-p ./complex/path/to/newService` or `-p newService`

cc @DavidWells

**Comment by flomotlik** [comment]
I like this. If necessary we can still add a `--name` attribute in the future if people want to create a project in the CWD but still give it a name. But imho its fine like this now.

**Comment by pmuens** [comment]
Alright. Then I'll update #1937 with @ac360 proposal

**Comment by pmuens** [comment]
@serverless/core just pushed an update to #1937 which implements the service renaming as describe above.

**Comment by flomotlik** [comment]
Merged
